File: import_build/src/Computation/dState.py
# //The class "dState" stores a differential of a
# // tangent vector to configuration space
# //in practice, this will the configuration space of a single ball system,
# //meaning that state.vel will be the ball's velocity and state.acc will be
# //covariant derivative with respect to the metric on the ambient space
# //so long as they implement the following methods:


# // .vel and .acc need to implement
# // .clone(), as well as the  vector space operations:
# // .add(), .sub() . multiplyScalar(), .applyMatrix3()

import numpy as np


class dState:
    """
    A class used to store information about dstate
    Describes vector

    ...

    Attributes
    ----------
    vel : array
        array of velocity data

    acc : array
        array of acceleration data

    Methods
    -------
    clone()
        Generate copy of self
        Returns dState clone

    add(dstate)
        Add velocity and acceleration of dstate to self.vel and self.acc, respectively
        Returns self

    sub(dstate)
        Subtract velocity and acceleration of dstate from self.vel and self.acc, respectively
        Returns self

    multiplyScalar(k)
        Scale self.vel and self.acc by scalar value k
        Returns self
    """

    # //build a dState from the input of an object storing velocity data
    # //and an object storing acceleration data
    def __init__(self, vel, acc):
        self.vel=vel.copy()
        self.acc=acc.copy()

    # //make a copy of a given dState (not just reference it in memory)
    # //and return the copy
    def clone(self):
        return  dState(self.vel.copy(),self.acc.copy())


    # //add the velocity AND of a given state to the current
    def add(self, dState):
        self.vel = np.add(self.vel, dState.vel)
        self.acc = np.add(self.acc, dState.acc)
        return self

    # //subtract the velocity AND acceleration of a given state from the current
    def sub(self, dState ):
        self.vel = np.subtract(self.vel, dState.vel )
        self.acc = np.subtract(self.acc, dState.acc )
        return self


    # //scale the velocity AND acceleration of the current state by a factor
    def multiplyScalar (self, k ):
        self.vel = self.vel * k
        self.acc = self.acc * k 
        return self

File: cpp_build/Computation/DState.cpp
#include <vector>
#include <iostream>
#include "../includes/eigen-3.4.0/Eigen/Dense"
#include "State.h"
#include "DState.h"


//build a dstate from the input of an object storing velocity data and an object storing acceleration data
DState::DState(Eigen::Vector3d vel, Eigen::Vector3d acc)
{
    this->vel = vel;
    this->acc = acc;
}

// DState::DState(std::vector<double> vel, std::vector<double> acc)
// {
//     this->vel = vel;
//     this->acc = acc;
// }

//make a copy of a given dstate (not just reference it in memory) and return the copy
DState DState::clone()
{
    return  DState(this->vel,this->acc);
}

//print information about the dstate to the terminal
void DState::print_info()
{
    
    std::cout << "DState Position: [" << this->vel(0) << " , " << this->vel(1) << " , " << this->vel(2) << " ]\n" << "DState Acceleration: [" << this->acc(0) << " , " << this->acc(1) << " , " << this->acc(2) << " ]\n";
    // std::cout << "DState Position: [" << this->vel[0] << " , " << this->vel[1] << " , " << this->vel[2] << " ]\n" << "DState Acceleration: [" << this->acc[0] << " , " << this->acc[1] << " , " << this->acc[2] << " ]\n";
}

//add the velocity AND acceleration of a given dstate to the current
void DState::add(DState dstate)
{ 
    // Add the elements of state.vec and this->vel
    this->vel = this->vel + dstate.vel;
    this->acc = this->acc + dstate.acc; 

}

// void DState::add(DState dstate)
// { 
//     // Add the elements of state.vec and this->vel
//     for (int i = 0; i < this->vel.size(); i++)  
//     {  
//         this->vel[i] = this->vel[i] + dstate.vel[i];
//         this->acc[i] = this->acc[i] + dstate.acc[i]; 
//     }
// }

//substract the velocity AND acceleration of a given dstate from the current
void DState::sub(DState dstate)
{
    // Subtract the elements of state.vel from this->vel
    this->vel = this->vel - dstate.vel;
    this->acc = this->acc - dstate.acc;

}

// void DState::sub(DState dstate)
// {
//     // Subtract the elements of state.vel from this->vel
//     for (int i = 0; i < this->vel.size(); i++)  
//     {  
//         this->vel[i] = this->vel[i] - dstate.vel[i];
//         this->acc[i] = this->acc[i] - dstate.acc[i]; 
//     }
// }

//scale the velocity AND acceleration of the current state by a factor
void DState::multiplyScalar(double k)
{
    // Multiply the elements of state.vel by k
    this->vel = this->vel * k; 
    this->acc = this->acc * k;  
}

// void DState::multiplyScalar(double k)
// {
//     // Multiply the elements of state.vel by k
//     for (int i = 0; i < this->vel.size(); i++)  
//     {  
//         this->vel[i] = this->vel[i] * k; 
//         this->acc[i] = this->acc[i] * k;  
//     }
// }
